R/maraca.R

```diff
@@ -347,124 +426,116 @@ plot_maraca <- function(
-  class(plot) <- c("maraca_tte_components", class(plot))
+  arms <- levels(pb$plot$data[, pb$plot$labels$colour])
 
-  return(plot)
-}
+  tte_data <- utils::tail(
+    utils::head(pb$data[[4]][, c("group", "x", "y")], -2), -2)
+  tte_data$group <- factor(tte_data$group, labels = arms)
 
-#' Generic to print the maraca_tte_component that is not a ggplot.
-#'
-#' @param x an object of S3 class 'maraca_tte_components'
-#' @param \dots not used
-#' @return Used for side effect. Plots the maraca_tte_components object.
-#'
-#' @examples
-#' data(hce_scenario_a)
-#' hce_test <- maraca(
-#'   data = hce_scenario_a,
-#'   tte_outcomes = c("Outcome I", "Outcome II", "Outcome III", "Outcome IV"),
-#'   continuous_outcome = "Continuous outcome",
-#'   column_names = c(outcome = "GROUP", arm = "TRTP", value = "AVAL0"),
-#'   arm_levels = c(active = "Active", control = "Control"),
-#'   compute_win_odds = TRUE
-#' )
-#' plot <- plot_tte_components(hce_test)
-#' print(plot)
-#'
-#' @export
-print.maraca_tte_components <- function(x, ...) {
-  grid::grid.draw(x)
-}
+  scatter_data <- NULL
+  boxstat_data <- NULL
+  violin_data <- NULL
 
+  if (plot_type == "GeomPoint") {
+    scatter_data <- pb$data[[5]][, c("group", "x", "y")]
+    scatter_data$group <- factor(scatter_data$group, labels = arms)
+  }
+  if (plot_type == "GeomBoxplot") {
+    boxstat_data <- pb$data[[5]] %>%
+      dplyr::select(.data$group, "x_lowest" = .data$xmin_final,
+                    "whisker_lower" = .data$xmin,
+             "hinge_lower" = .data$xlower, "median" = .data$xmiddle,
+             "hinge_upper" = .data$xupper, "whisker_upper" = .data$xmax,
+             "x_highest" = .data$xmax_final, .data$outliers)
+    boxstat_data$outliers <- lapply(boxstat_data$outliers, sort)
+    boxstat_data$group <- factor(boxstat_data$group, labels = arms)
+  }
+  if (plot_type == "GeomViolin") {
+    violin_data <- pb$data[[5]][, c("group", "x", "y", "density", "width")]
+    violin_data$group <- factor(violin_data$group, labels = arms)
+    if (class(as.list(x$layers[[6]])[["geom"]])[1] == "GeomBoxplot") {
+      boxstat_data <- pb$data[[6]] %>%
+        dplyr::select(.data$group, "x_lowest" = .data$xmin_final,
+                      "whisker_lower" = .data$xmin,
+               "hinge_lower" = .data$xlower, "median" = .data$xmiddle,
+               "hinge_upper" = .data$xupper, "whisker_upper" = .data$xmax,
+               "x_highest" = .data$xmax_final, .data$outliers)
+      boxstat_data$outliers <- lapply(boxstat_data$outliers, sort)
+      boxstat_data$group <- factor(boxstat_data$group, labels = arms)
+    }
+  }
+
+  if ("win_odds_parameters" %in% names(attributes(x))) {
+    params <- attr(x, "win_odds_parameters")
+    wo_stats <- c(winodds = params$win_odds,
+                  lowerCI = params$lower_ci,
+                  upperCI = params$upper_ci,
+                  p_value = params$p_value)
+  } else {
+    wo_stats <- "Not calculated"
+  }
+
+  return(
```

Comment:
Nice idea. However remember that the plotting can still fail. The renderer may break down (and it does, often), so there's no escape from checking visually a plot. Of course this is a good first defense check, but you still have to visually verify.

Reply:
Yes, this function is actually specifically added for a publication that compares the maraca plot from the package with another one done in SAS. There will be a visual comparison as well but this is allowing to also programmatically compare.